File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0;
        int maxlen=0;
        HashMap<Character,Integer> freq=new HashMap<>();
        for(int r=0;r<s.length();r++){
            char ch=s.charAt(r);
            if(freq.containsKey(ch)&& freq.get(ch)>=l){
                l=freq.get(ch)+1;
            }
            freq.put(ch,r);
            maxlen=Math.max(maxlen,r-l+1);


        }
        return maxlen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        kick=set()
        i=0
        maxlen=0
        for j in range(len(s)):
            while(s[j] in kick):
                kick.remove(s[i])
                i+=1
            kick.add(s[j])
            maxlen=max(maxlen,j-i+1)
        return  maxlen
            
